### PR TITLE
Interface Approach2

### PR DESCRIPTION
**Pull request by JuiP**
Work in progress.

- [x] Split ips and snips into two files
- [x] Split code into two classes: Estimator and Interval
- [x] Split pseudo_inverse code into two classes: Estimator and Interval
- [x] Adding code for interfaces

### CODE REVIEW

**Comment by JuiP** [comment]
@cheng-tan @ataymano :smile: Please have a look...

There are 2 things we might want to discuss:

- [ ] Instead of returning nothing in add_example() how about we return `self.data`. For someone executing the `get_estimate()` followed by `get_interval()` can make store the data returned by `add_example()`. By doing this, we can save executing the same function on the same data for getting the CI. (*side note*: Also discuss if unused dictionary keys are required under this method)

- [ ] For the Contextual Bandits interface, the `get_interval()` method for cressieread has 4 parameters:
``` def get_interval(self, alpha=0.05, rmin=0, rmax=1): ```
But for snips and ips the `get_interval()` method has 3 parameters:
``` def get_interval(self, type, alpha=0.05): ```

File: estimators/contextual_bandits/cb_helper.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+""" Interface for implementation of slates estimator """
+
+from abc import ABC, abstractmethod
+from typing import List
+
+class CbEstimator(ABC):
```
Comment (ataymano) [inline estimators/contextual_bandits/cb_helper.py:6]: we are already inside contextual_bandit context, so Cb prefix seems redundant:
contextual_bandit.CbEstimator sounds like contextual_bandit.SlatesEstgimator could exist as well, which is confusing.
Maybe we can just have Estimator / Interval?

File: estimators/contextual_bandits/cb_helper.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
+""" Interface for implementation of slates estimator """
```
Comment (cheng-tan) [inline estimators/contextual_bandits/cb_helper.py:1]: this comment seems misleading, this is the cb estimator

File: estimators/contextual_bandits/cb_helper.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+from abc import ABC, abstractmethod
+from typing import List
+
+class CbEstimator(ABC):
+	""" Interface for implementation of slates estimator """
```
Comment (cheng-tan) [inline estimators/contextual_bandits/cb_helper.py:7]: also this one.

File: estimators/contextual_bandits/cb_helper.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	def get_estimate(self) -> float:
+		""" Calculates the selected estimator
+		
+		Returns:
+			The estimator value
+		"""
+		...
+
+class CbInterval(ABC):
+	""" Interface for implementation of slates estimator interval """
```
Comment (cheng-tan) [inline estimators/contextual_bandits/cb_helper.py:30]: same

File: estimators/contextual_bandits/cb_helper.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+		Args:
+			p_log: probability
+			r: reward value
+			p_pred:
+			count:
+		"""
+		...
+
+	@abstractmethod
+	def get_interval(self, type: str, alpha: float) -> List:
```
Comment (cheng-tan) [inline estimators/contextual_bandits/cb_helper.py:44]: we don't want to pass `type` as an arg, only alpha

File: estimators/contextual_bandits/cb_helper.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
+""" Interface for implementation of slates estimator """
+
```
Comment (cheng-tan) [inline estimators/contextual_bandits/cb_helper.py:2]: Also this file name makes me think it includes util functions. Maybe rename it to base.py?

File: estimators/contextual_bandits/cressieread.py [hunk 7]
```diff
@@ -1,8 +1,9 @@
 # CR(-2) is particularly computationally convenient
 
 from math import fsum, inf
+from cb_helper import CbEstimator
 
-class Estimator:
+class Estimator(CbEstimator):
```
Comment (cheng-tan) [inline estimators/contextual_bandits/cressieread.py:6]: is we rename the CBEstimator to Estimator, this one can be CressiereadEstimator(Estimator)

File: estimators/contextual_bandits/cressieread.py [hunk 8]
```diff
@@ -24,7 +25,7 @@ def add_example(self, p_log, r, p_pred, count=1):
             self.wmax = max(self.wmax, w)
             self.wmin = min(self.wmin, w)
 
-    def get_estimate(self, rmin=0, rmax=1):
+    def get_estimate(self):
```
Comment (cheng-tan) [inline estimators/contextual_bandits/cressieread.py:28]: We need to add code to calculate `rmin` & `rmax`

File: estimators/contextual_bandits/cb_helper.py (hunk 3 above)
Comment (cheng-tan) [inline estimators/contextual_bandits/cb_helper.py:7]: Do we need a __init__() method?

File: estimators/contextual_bandits/cb_helper.py (hunk 2 above)
Comment (ataymano) [inline estimators/contextual_bandits/cb_helper.py:1]: typo in comment

File: estimators/contextual_bandits/ips.py [hunk 9]
```diff
@@ -71,4 +79,4 @@ def get_interval(self, type, alpha=0.05):
 
         if not bounds:
             bounds = [0, 0]
-        return bounds
+        return bounds
```
Comment (cheng-tan) [inline estimators/contextual_bandits/ips.py:82]: add a newline at the end of the file

File: estimators/contextual_bandits/mle.py [hunk 10]
```diff
@@ -28,7 +29,7 @@ def graddualobjective(self, n, beta):
        return fsum(c * (w - 1)/((w - 1) * beta + n)
                   for c, w, _ in self.data)
 
-    def get_estimate(self, rmin=0, rmax=1):
+    def get_estimate(self):
```
Comment (cheng-tan) [inline estimators/contextual_bandits/mle.py:32]: same here need to calculate rmin and rmax

File: estimators/contextual_bandits/snips.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    def add_example(self, p_log, r, p_pred, count=1):
+        self.data['N'] += count
+        if p_pred > 0:
+            p_over_p = p_pred/p_log
+            if r != 0:
+                self.data['n'] += r*p_over_p*count
+                self.data['c'] = max(self.data['c'], r*p_over_p)
+                self.data['SoS'] += ((r*p_over_p)**2)*count
+
+    def get_interval(self, type, alpha=0.05):
```
Comment (cheng-tan) [inline estimators/contextual_bandits/snips.py:59]: for this one we don't need to pass in type, but have `GaussianInterval(Interval)` and `ClopperPearsonInterval(Interval)`

File: estimators/slates/slates_helper.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
+""" Interface for implementation of slates estimator """
+
+from abc import ABC, abstractmethod
+from typing import List
+
+class SlatesEstimator(ABC):
+	""" Interface for implementation of slates estimator """
+
+	@abstractmethod
+	def add_example(self, p_log: float, r: float, p_pred: float, count: int) -> None:
```
Comment (cheng-tan) [inline estimators/slates/slates_helper.py:10]: So this one, we need to discuss a bit more, slates p_log and p_pred are arrays not float

File: estimators/slates/slates_helper.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+		Returns:
+			The estimator value
+		"""
+		...
+
+class SlatesInterval(ABC):
+	""" Interface for implementation of slates estimator interval """
+
+	@abstractmethod
+	def add_example(self, p_log: float, r: float, p_pred: float, count: int) -> None:
```
Comment (cheng-tan) [inline estimators/slates/slates_helper.py:34]: same here

File: estimators/slates/slates_helper.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+		Args:
+			p_log: probability
+			r: reward value
+			p_pred:
+			count:
+		"""
+		...
+
+	@abstractmethod
+	def get_interval(self, type: str, alpha: float) -> List:
```
Comment (cheng-tan) [inline estimators/slates/slates_helper.py:45]: Also shouldn't pass in type

**Review by cheng-tan — COMMENTED** [review]
Also we need to iterate on the doc of the args.

File: estimators/contextual_bandits/cb_helper.py (hunk 1 above)
Comment (ataymano) [reply estimators/contextual_bandits/cb_helper.py:6]: File name is a bit strange - "*_helper" is more appropriate for some secondary stuff that is not 100% relevant to the main purpose of the project. Probably here something like "core" would sound better.

**Comment by cheng-tan** [comment]
Hi @JuiP this is the updated structure we discussed today, in case you need to reference:
- estimators
  - init
  - basic_usage
  - cb
    - init
    - base.py / core.py
    - cat_utils
    - cressieread
    - mle
    - ips
    - snips
    - gaussian
    - clopper_pearson
  - slates
    - pi
    - gaussian
  - utils
    - ds_parse
  - test

File: estimators/contextual_bandits/cb_helper.py (hunk 2 above)
Comment (JuiP) [reply estimators/contextual_bandits/cb_helper.py]: Changed in 3a27447

Comment (JuiP) [reply estimators/contextual_bandits/cb_helper.py]: Changed in 3a27447

File: estimators/contextual_bandits/cb_helper.py (hunk 3 above)
Comment (JuiP) [reply estimators/contextual_bandits/cb_helper.py]: Changed in 3a27447

File: estimators/contextual_bandits/cb_helper.py (hunk 4 above)
Comment (JuiP) [reply estimators/contextual_bandits/cb_helper.py]: Changed in 3a27447

File: estimators/contextual_bandits/cb_helper.py (hunk 6 above)
Comment (JuiP) [reply estimators/contextual_bandits/cb_helper.py]: Done in 045d894

File: estimators/contextual_bandits/cressieread.py (hunk 7 above)
Comment (JuiP) [reply estimators/contextual_bandits/cressieread.py]: Done in 045d894

File: estimators/contextual_bandits/cb_helper.py (hunk 1 above)
Comment (JuiP) [reply estimators/contextual_bandits/cb_helper.py]: Sure I have made the changes in 045d894.

**Comment by JuiP** [comment]
@ataymano @cheng-tan 
I was looking at changing `count` to `weight` as discussed in the meet... I just came across [this in Cressieread](https://github.com/JuiP/estimators/blob/7567130e438e78b8dbac7f6b3fc9093b73553be5/estimators/contextual_bandits/cressieread.py#L21) and [this in ips](https://github.com/JuiP/estimators/blob/7567130e438e78b8dbac7f6b3fc9093b73553be5/estimators/contextual_bandits/ips.py#L48). I have two questions:

- Since in both cases the same thing is being calculated, for consistency should I change `p_over_p` to `w` or vice-versa?
- Will changing `count` to `weight` be confusing?

File: estimators/contextual_bandits/cb_helper.py (hunk 3 above)
Comment (JuiP) [reply estimators/contextual_bandits/cb_helper.py]: I'm not sure what would the **init()** method contain for Contextual bandits? For *cressieread* `self.data` is a list and for *ips* it is a dictionary... Apart from that `self.wmin` and `self.wmax` is being calculated only for cressieread and mle.

Maybe we can add the **init()** method for slates interface?

File: estimators/contextual_bandits/snips.py (hunk 11 above)
Comment (JuiP) [reply estimators/contextual_bandits/snips.py]: Done in 4f3a29f

File: estimators/contextual_bandits/cb_helper.py (hunk 5 above)
Comment (JuiP) [reply estimators/contextual_bandits/cb_helper.py]: Changed in 4f3a29f

File: estimators/contextual_bandits/ips.py (hunk 9 above)
Comment (JuiP) [reply estimators/contextual_bandits/ips.py]: Done in b023d1e

File: estimators/slates/slates_helper.py (hunk 14 above)
Comment (JuiP) [reply estimators/slates/slates_helper.py]: Changed in 53c9b7a

File: estimators/slates/slates_helper.py (hunk 13 above)
Comment (JuiP) [reply estimators/slates/slates_helper.py]: Changed in 53c9b7a

File: estimators/slates/slates_helper.py (hunk 12 above)
Comment (JuiP) [reply estimators/slates/slates_helper.py]: Changed in 53c9b7a

**Comment by JuiP** [comment]
I feel I shouldn't have made any change to the file `test/test_pi.py` in this PR. This PR has a lot of changes already. So should I leave it as it is or fix it here? :sweat_smile:

File: estimators/contextual_bandits/cb_base.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
+""" Interface for implementation of contextual bandit estimators """
```
Comment (ataymano) [inline estimators/contextual_bandits/cb_base.py:1]: Seems like contextual_bandits.cb_base is mentioning "contextual bandits" concept twice which seems redundat (as "contextual_bandits" and as "cb"). Also, probably there is no chance to have something like contextual_bandits.slates_base, so it seems reasonable to rename cb_base to base (and similarly for slates)

File: estimators/contextual_bandits/ips.py [hunk 16]
```diff
@@ -0,0 +1,26 @@
+from cb_base import Estimator
+
+class IPSestimator(Estimator):
```
Comment (ataymano) [inline estimators/contextual_bandits/ips.py:3]: Looks like there is inconsistency about the way we use "estimator" suffix ("Estimator" vs "estimator"). Can we make it consistent?

**Comment by ataymano** [comment]
> @ataymano @cheng-tan
> I was looking at changing `count` to `weight` as discussed in the meet... I just came across [this in Cressieread](https://github.com/JuiP/estimators/blob/7567130e438e78b8dbac7f6b3fc9093b73553be5/estimators/contextual_bandits/cressieread.py#L21) and [this in ips](https://github.com/JuiP/estimators/blob/7567130e438e78b8dbac7f6b3fc9093b73553be5/estimators/contextual_bandits/ips.py#L48). I have two questions:
> 
> * Since in both cases the same thing is being calculated, for consistency should I change `p_over_p` to `w` or vice-versa?
> * Will changing `count` to `weight` be confusing?

File: estimators/contextual_bandits/ips.py (hunk 16 above)
Comment (JuiP) [reply estimators/contextual_bandits/ips.py:3]: Yes, I wasn't sure if I should go with `IPSEstimator` or IpsEstimator`...

**Comment by ataymano** [comment]
> I feel I shouldn't have made any change to the file `test/test_pi.py` in this PR. This PR has a lot of changes already. So should I leave it as it is or fix it here? 😅

Is it just renaming get_estimate to get? If so, it is probably to do it in this PR since we already have these lines change in the diff

File: estimators/contextual_bandits/cb_base.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+		Args:
+			p_log: probability of the logging policy
+			r: reward for choosing an action in the given context
+			p_pred: predicted probability of making decision
+			count: weight
+		"""
+		...
+
+	@abstractmethod
+	def get(self, alpha: float) -> List:
```
Comment (ataymano) [inline estimators/contextual_bandits/cb_base.py:44]: to consider for next PRs: List -> tuple

Comment (ataymano) [reply estimators/contextual_bandits/cb_base.py:44]: to consider for next PRs: default value for alpha

File: estimators/contextual_bandits/cb_base.py [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+		""" Calculates the CI
+		Args:
+			alpha: alpha value
+		Returns:
+			Returns the confidence interval as list[float]
```
Comment (cheng-tan) [inline estimators/contextual_bandits/cb_base.py:49]: for next pr: update the doc to be list/tuple of min_bound and max_bound

File: estimators/contextual_bandits/clopper_pearson.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+            p_over_p = p_pred/p_log
+            if r != 0:
+                self.data['n'] += r*p_over_p*count
+                self.data['c'] = max(self.data['c'], r*p_over_p)
+
+    def get(self, alpha=0.05):
+        bounds = []
+        num = self.data['n']
+        den = self.data['N']
+        maxWeightedCost = self.data['c']
```
Comment (cheng-tan) [inline estimators/contextual_bandits/clopper_pearson.py:30]: variable naming is also inconsistent. should be fixed in next prs.

File: estimators/slates/slates_base.py [hunk 20]
```diff
@@ -0,0 +1,50 @@
+""" Interface for implementation of slates estimator """
+
+from abc import ABC, abstractmethod
+from typing import List
+
+class Estimator(ABC):
+	""" Interface for implementation of slates estimator """
+
+	@abstractmethod
+	def add_example(self, p_log: List, r: float, p_pred: List, count: float) -> None:
```
Comment (cheng-tan) [inline estimators/slates/slates_base.py:10]: p_logs and p_preds ?

File: estimators/slates/slates_base.py [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+		""" 
+		Args:
+			p_log: probability of the logging policy
```
Comment (cheng-tan) [inline estimators/slates/slates_base.py:13]: Also update the doc here to indicate p_log and p_pred should be list

File: estimators/test/test_pi.py [hunk 22]
```diff
@@ -2,13 +2,14 @@
...
-from contextual_bandits import ips_snips
+from contextual_bandits import ips
+from contextual_bandits import snips
 from contextual_bandits import cats_utils
 
 def test_single_slot_pi_equivalent_to_ips():
     """PI should be equivalent to IPS when there is only a single slot"""
     pi_estimator = pseudo_inverse.Estimator()
-    ips_estimator = ips_snips.Estimator()
+    ips_estimator = ips.Estimator()
```
Comment (cheng-tan) [inline estimators/test/test_pi.py:12]: Let's update test_pi with this PR, at least make the existing tests pass.

File: estimators/contextual_bandits/cb_base.py (hunk 15 above)
Comment (JuiP) [reply estimators/contextual_bandits/cb_base.py]: Sure! Done here: 1429f71

File: estimators/contextual_bandits/ips.py (hunk 16 above)
Comment (JuiP) [reply estimators/contextual_bandits/ips.py]: Done here: 1429f71

File: estimators/slates/slates_base.py (hunk 20 above)
Comment (JuiP) [reply estimators/slates/slates_base.py]: Ohh yes, sorry I skipped this one. Done here: 813ddc3

File: estimators/slates/slates_base.py (hunk 21 above)
Comment (JuiP) [reply estimators/slates/slates_base.py]: Done here: 813ddc3

File: estimators/test/test_pi.py (hunk 22 above)
Comment (JuiP) [reply estimators/test/test_pi.py]: Definitely! I have made the required changes here: c797fbf

**Comment by JuiP** [comment]
I think there is just one change left: Cressieread Interval class to use the interface. Then its good to go :rocket:
Do let me know if there are more changes required for merging this PR :smile:

File: estimators/test/test_pi.py [hunk 23]
```diff
@@ -2,13 +2,14 @@
...
-from contextual_bandits import ips_snips
+from contextual_bandits import ips
+from contextual_bandits import snips
 from contextual_bandits import cats_utils
 
 def test_single_slot_pi_equivalent_to_ips():
     """PI should be equivalent to IPS when there is only a single slot"""
-    pi_estimator = pseudo_inverse.Estimator()
-    ips_estimator = ips_snips.Estimator()
+    pi_estimator = pseudo_inverse.PseudoInverseEstimator()
```
Comment (cheng-tan) [inline estimators/test/test_pi.py:11]: this seems a bit repetitive.

**Review by cheng-tan — APPROVED** [review]

File: estimators/test/test_pi.py (hunk 23 above)
Comment (ataymano) [reply estimators/test/test_pi.py:11]: ok, probably it is better to change it to pseudo_inverse.Estimator (and similarly for everything else)

File: estimators/contextual_bandits/cressieread.py [hunk 24]
```diff
@@ -53,17 +55,20 @@ def get_estimate(self, rmin=0, rmax=1):
...
 
         self.wmin = wmin
         self.wmax = wmax
 
+        self.rmin = rmin
+        self.rmax = rmax
+
         self.data = []
 
     def add_example(self, p_log, r, p_pred, count=1):
```
Comment (ataymano) [inline estimators/contextual_bandits/cressieread.py:74]: todo: what to do if r is out of [rmin, rmax]?

**Review by ataymano — APPROVED** [review]

**Pull request merged**